src/uharfbuzz/_harfbuzz.pyx

```diff
@@ -579,34 +579,50 @@ cdef const char ** to_cstring_array(list_str):
     return ret
 
 
-def shape(font: Font, buffer: Buffer, features: Dict[str, bool] = None,
+def shape(font: Font, buffer: Buffer,
+        features: Dict[str,Union[int,bool,Sequence[Tuple[int,int,Union[int,bool]]]]] = None,
         shapers: List[str] = None) -> None:
     cdef unsigned int size
     cdef hb_feature_t* hb_features
     cdef bytes packed
-    cdef char* cstr
     cdef hb_feature_t feat
     cdef const char **c_shapers
-    if features is None:
-        size = 0
-        hb_features = NULL
-    else:
-        size = len(features)
-        hb_features = <hb_feature_t*>malloc(size * sizeof(hb_feature_t))
-        for i, (name, value) in enumerate(features.items()):
-            packed = name.encode()
-            cstr = packed
-            hb_feature_from_string(packed, len(packed), &feat)
-            feat.value = value
-            hb_features[i] = feat
-    if shapers:
-        c_shapers = to_cstring_array(shapers)
-        hb_shape_full(font._hb_font, buffer._hb_buffer, hb_features, size, c_shapers)
-        free(c_shapers)
-    else:
-        hb_shape(font._hb_font, buffer._hb_buffer, hb_features, size)
-    if hb_features is not NULL:
-        free(hb_features)
+    size = 0
+    hb_features = NULL
+    try:
+        if features:
+            for value in features.values():
+                if isinstance(value, int):
+                    size += 1
+                else:
+                    size += len(value)
+            hb_features = <hb_feature_t*>malloc(size * sizeof(hb_feature_t))
+            i = 0
+            for name, value in features.items():
+                assert i < size, "index out of range for feature array capacity"
+                packed = name.encode()
+                if isinstance(value, int):
+                    hb_feature_from_string(packed, len(packed), &feat)
```

Comment:
Please change this to `hb_tag_from_string()` like below as well.

Reply:
That would make this change non-backwards compatible, are we ok with that?

Reply:
Intresting. *If* we concluded the range numbers work fine (which looks like they do), then I'm agree keeping it is fine.

Reply:
Not sure what you mean. I think the range numbers are fine.

But before this PR, the `features` dict already supported a notation like `features={"aalt[2:3]", 2}`. That feels accidental to me (I didn't even realized it did that until Khaled pointed it out), but perhaps @adrientetar built it this way on purpose, and perhaps there are users out there that use it.

Reply:
Yes yes. I understand. And I said I agree with your assessment. lgtm.